pkg/services/hivelocity/device/device.go

```diff
@@ -393,3 +343,134 @@ func createTags(clusterName, machineName string, isControlPlane bool) []string {
 		"caphv-machinetype=" + machineType,
 	}
 }
+
+// setMachineAddress gets the address from the device and sets it on the Machine object.
+func setMachineAddress(hvMachine *infrav1.HivelocityMachine, hvDevice *hv.BareMetalDevice) {
+	hvMachine.Status.Addresses = []clusterv1.MachineAddress{
+		{
+			Type:    clusterv1.MachineHostName,
+			Address: hvDevice.Hostname,
+		},
+		{
+			Type:    clusterv1.MachineInternalIP,
+			Address: hvDevice.PrimaryIp,
+		},
+		{
+			Type:    clusterv1.MachineExternalIP,
+			Address: hvDevice.PrimaryIp,
+		},
+	}
+}
+
+// associateDevice claims an unused HV device.
+func (s *Service) associateDevice(ctx context.Context, device *hv.BareMetalDevice) error {
+	deviceTags := append(device.Tags, clusterAndMachineTag(s.scope.HivelocityCluster.Name, s.scope.Name())...)
+	if err := s.scope.HVClient.SetTags(ctx, device.DeviceId, deviceTags); err != nil {
+		return fmt.Errorf("failed to set tags on machine %s :%w", s.scope.Name(), err)
+	}
+	// Set update tags of device to have the latest device for the next steps in reconcile
+	device.Tags = deviceTags
+	return nil
+}
+
+func clusterAndMachineTag(clusterName, machineName string) []string {
+	return []string{
+		hvclient.TagKeyClusterName + "=" + clusterName,
+		hvclient.TagKeyMachineName + "=" + machineName,
+	}
+}
+
+// chooseDevice search for an unused device, and then associates the device.
+func (s *Service) chooseDevice(ctx context.Context) (*hv.BareMetalDevice, error) {
+	devices, err := s.scope.HVClient.ListDevices(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("[chooseDevice] ListDevices() failed. machine %q: %w",
+			s.scope.Name(), err)
+	}
+	return chooseAvailableFromList(devices, s.scope.HivelocityMachine.Spec.Type, s.scope.Cluster.Name)
+}
+
+func chooseAvailableFromList(devices []*hv.BareMetalDevice, deviceType infrav1.HivelocityDeviceType, clusterName string) (*hv.BareMetalDevice, error) {
+	for _, device := range devices {
+		// Ignore if associated already
+		if isAssociated(device) {
+			continue
+		}
+
+		// Ignore if has wrong device type
+		dt, err := getDeviceType(device)
+		if err != nil {
+			if errors.Is(err, ErrNoMatchingTagFound) {
+				continue
+			}
+			return nil, fmt.Errorf("[chooseAvailableFromList] getDeviceType() failed: %w", err)
```

Comment:
I guess we shouldn't return an error here if too many tags have been found. We can maybe emit a log or whatever, we can talk about it. But in this way, all clusters would fail if someone misconfigured one HV device.

Reply:
Good hint. I updated the code. Errors get logged now, and then a "continue" gets done.